python/lsst/daf/butler/registry/queries/expressions/convert.py

```diff
@@ -892,7 +909,11 @@ def visitStringLiteral(self, value: str, node: Node) -> WhereClauseConverter:
 
     def visitTimeLiteral(self, value: Time, node: Node) -> WhereClauseConverter:
         # Docstring inherited from TreeVisitor.visitTimeLiteral
-        return ScalarWhereClauseConverter.fromLiteral(value)
+        if node.user_data is datetime:
+            dt = value.to_datetime()
+            return ScalarWhereClauseConverter.fromLiteral(dt)
+        else:
+            return ScalarWhereClauseConverter.fromLiteral(value)
```

Comment:
I think the approach you've taken here is fine, but I'll mention another one here in case you like it better:
 - Make a new `TimeLiteralWhereClauseConverter` subclass for time literals that only stores the `str` from the expression, and does not coerce it to either `astropy.time.Time`, but still declares its `dtype` to be `astropy.time.Time`.
 - Remove `Time` from the `registerBinary` calls for other comparison operators, and instead register `Time`-only ones that delegate to a func that coerces those literals according to the other operand.

I think that generalizes better to other kinds of context-dependent coercion, though of course it'd be nice if we didn't need any more of that.

Reply:
@TallJimbo, thanks for suggestion! I have decided to simplify things based on that idea and remove two-pass tree analysis in favor of doing what you suggest based on type of operands in comparison operators. This greatly simplifies the whole thing but with a caveat of not supporting some complex expressions of a kind `ingest_date < MAX(time_literal1, time_literal2)` but we do not currently support this syntax anyways. I think we'll be OK with this for now, and if we need to extend our language support for more complex expressions we'll need to rethink how the whole conversion works (or do something with Time/datetime duality).

One thing that I noticed is that we do not handle timestaps in `IN` operator, even though documentation says it should work. I'll open a new ticket for that.

Could you look at it once again, should be shorter commit this time.